File: reference-implementations/validator-node/src/loader.ts
// Manifest + sibling-manifest loader. Kept thin on purpose. YAML parsing is
// delegated to the `yaml` package; everything else operates on plain objects.

import { readFileSync } from "node:fs";
import * as path from "node:path";
import { parse as parseYaml } from "yaml";
import { globSync } from "glob";

export type Manifest = Record<string, unknown>;

export function loadYaml(filePath: string): Manifest {
  const raw = readFileSync(filePath, "utf-8");
  const doc = parseYaml(raw);
  if (doc == null || typeof doc !== "object" || Array.isArray(doc)) {
    return {};
  }
  return doc as Manifest;
}

export function findSiblingManifests(
  manifestPath: string
): Record<string, Manifest> {
  const dir = path.dirname(path.resolve(manifestPath));
  const self = path.resolve(manifestPath);
  const siblings: Record<string, Manifest> = {};
  const candidates = globSync("change-manifest*.yaml", { cwd: dir }).sort();
  for (const name of candidates) {
    const abs = path.resolve(dir, name);
    if (abs === self) continue;
    let doc: Manifest;
    try {
      doc = loadYaml(abs);
    } catch {
      continue;
    }
    const cid = doc["change_id"];
    if (typeof cid === "string" && cid.length > 0) {
      siblings[cid] = doc;
    }
  }
  return siblings;
}

File: reference-implementations/validator-python/src/agent_protocol_validate/loader.py
"""Manifest + sibling-manifest loader.

Kept thin on purpose. YAML parsing is delegated to PyYAML; everything else
(schema, graph building, drift) operates on plain dicts.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

import yaml


def load_yaml(path: Path) -> dict[str, Any]:
    with path.open("r", encoding="utf-8") as fh:
        return yaml.safe_load(fh) or {}


def find_sibling_manifests(manifest_path: Path) -> dict[str, dict[str, Any]]:
    """Return ``{change_id: manifest_dict}`` for every other manifest in
    the same directory. Used for the 2.5 bidirectional-mirror check.
    """
    siblings: dict[str, dict[str, Any]] = {}
    for candidate in sorted(manifest_path.parent.glob("change-manifest*.yaml")):
        if candidate.resolve() == manifest_path.resolve():
            continue
        try:
            data = load_yaml(candidate)
        except yaml.YAMLError:
            continue
        change_id = data.get("change_id")
        if isinstance(change_id, str) and change_id:
            siblings[change_id] = data
    return siblings
